src/main/java/de/tum/in/www1/artemis/service/ParticipationVCSAccessTokenService.java

```diff
@@ -0,0 +1,90 @@
+import de.tum.in.www1.artemis.repository.ParticipationVCSAccessTokenRepository;
+import de.tum.in.www1.artemis.service.connectors.localvc.LocalVCPersonalAccessTokenManagementService;
+
+@Profile(PROFILE_CORE)
+@Service
+public class ParticipationVCSAccessTokenService {
+
+    private static final Logger log = LoggerFactory.getLogger(ParticipationVCSAccessTokenService.class);
+
+    private final ParticipationVCSAccessTokenRepository participationVCSAccessTokenRepository;
+
+    public ParticipationVCSAccessTokenService(ParticipationVCSAccessTokenRepository participationVCSAccessTokenRepository) {
+        this.participationVCSAccessTokenRepository = participationVCSAccessTokenRepository;
+    }
+
+    /**
+     * Creates a vcs access token for a User,Participation pair and stores it in the database
+     *
+     * @param user          the user which is owner of the token
+     * @param participation the participation which belongs to the token
+     */
+    public void createParticipationVCSAccessToken(User user, StudentParticipation participation) {
+        ParticipationVCSAccessToken participationVCSAccessToken = new ParticipationVCSAccessToken();
+        participationVCSAccessToken.setUser(user);
+        participationVCSAccessToken.setParticipation(participation);
+        participationVCSAccessToken.setVcsAccessToken(LocalVCPersonalAccessTokenManagementService.generateSecureVCSAccessToken());
+        participationVCSAccessTokenRepository.save(participationVCSAccessToken);
+    }
+
+    /**
+     * Creates vcs tokens for StudentParticipations which do not yet have one
+     *
+     * @param participations The participations, for which tokens should be generated, if they do not already have one
+     */
+    public void createMissingParticipationVCSAccessTokens(List<ProgrammingExerciseStudentParticipation> participations) {
+        var existingTokens = participationVCSAccessTokenRepository.findAllByParticipationIds(participations.stream().map(DomainObject::getId).toList());
+        log.debug("Create missing VcsAccessTokens for participationIds: {}", participations);
+        List<ParticipationVCSAccessToken> vcsAccessTokens = participations.stream()
+                .filter(participation -> !existingTokens.stream().map(ParticipationVCSAccessToken::getParticipation).toList().contains(participation)
+                        && participation.getParticipant() instanceof User)
+                .map(participation -> {
+                    var participationVCSAccessToken = new ParticipationVCSAccessToken();
+                    participationVCSAccessToken.setUser((User) participation.getParticipant());
+                    participationVCSAccessToken.setParticipation(participation);
+                    participationVCSAccessToken.setVcsAccessToken(LocalVCPersonalAccessTokenManagementService.generateSecureVCSAccessToken());
+                    return participationVCSAccessToken;
```

Comment:
Returns in Lambdas are often hard to read, I'd suggest to replace the lambda with a normal for loop.

Reply:
That's a good idea, that made it simpler and clearer to read :+1: